### DISCUSSION TITLE
How to inject "use client" directive

### DISCUSSION THREAD

**Discussion by wereHamster** [mdx-js/mdx#2570] [Q&A]
Is there a way to inject a "use client" directive at the top of the module that mdx-js compiler generates? Either via something in the .mdx file itself or via a plugin.

The "use client" directive is required when the `providerImportSource` option points to a React hook (which are not allowed in React Server Components).

**Comment by remcohaszing** [comment]
To directly address your question: Yes, you could definitely create a recma plugin that injects the `"use client"` directive.

The counter question is: Should you? MDX files typically contain a lot of text content that benefits from SSR.

This likely comes up, because you’re pointing `providerImportSource` to `@mdx-js/react` in a Next.js application. You should use `mdx-components.ts` instead.

**Comment by n9** [comment]
@remcohaszing My use case is to pass data to a MDX component from a react context.

UPDATE: This seems to be sufficient (for simple cases):

```typescript
import type { Program } from "estree";

export default function recmaUseClient() {
    return async function (tree: Program): Promise<Program> {
        tree.body.unshift({
            type: "ExpressionStatement",
            expression: {
                type: "Literal",
                value: "use client",
            },
            directive: "use client"
        })

        return tree
    }
}
```

**Reply by remcohaszing** [reply]
That seems about right. My only tips would be that you don’t have to return the tree, and the plugin can be made synchronous.

My main concern from [my last comment](https://github.com/orgs/mdx-js/discussions/2570#discussioncomment-11591381) still stands: Should you?

This plugin makes every MDX file a client component. I imagine you have some MDX like this:

```mdx
{/* page.mdx */}
import { use } from 'react'
import { UserContext } from './user.tsx'

# Hello {use(UserContext).name}!
```

But you could rewrite this to:

```tsx
// user-name.tsx
'use client'

import type { ReactNode } from 'react'
import { use } from 'react'
import { UserContext } from './user.tsx'

export function UserName(): ReactNode {
  return use(UserContext).name
}
```

```mdx
{/* page.mdx */}
import { UserName } from './user-name.tsx'

# Hello <UserName />!
```

Now you only need to hydrate the `UserName` component, not the entire MDX file.
